File: src/cdp_browser/perplexity.py
"""Perplexity API fallback for web research.

Used when CDP browser is unavailable or unreliable.
Requires PERPLEXITY_API_KEY in .env
"""
import os
import httpx
from dotenv import load_dotenv

load_dotenv()

PPLX_URL = "https://api.perplexity.ai/chat/completions"
PPLX_MODEL = "sonar"


def web_research(query: str) -> str:
    """Search the web via Perplexity API.

    Returns a web-grounded answer as plain text.
    No browser required — uses PERPLEXITY_API_KEY from .env.
    """
    api_key = os.getenv("PERPLEXITY_API_KEY")
    if not api_key:
        raise EnvironmentError(
            "PERPLEXITY_API_KEY not set in .env"
        )

    resp = httpx.post(
        PPLX_URL,
        headers={"Authorization": f"Bearer {api_key}"},
        json={
            "model": PPLX_MODEL,
            "messages": [
                {"role": "user", "content": query}
            ],
        },
        timeout=30,
    )
    resp.raise_for_status()
    data = resp.json()
    return (
        data["choices"][0]["message"]["content"]
    )

File: server_auto/src/cdp_browser/perplexity.rs
/// Perplexity API fallback for web research (no browser required).
use anyhow::{bail, Result};
use common::config;

const PPLX_URL: &str = "https://api.perplexity.ai/chat/completions";
const PPLX_MODEL: &str = "sonar";

pub async fn web_research(query: &str) -> Result<String> {
    let api_key = &config::get().perplexity_api_key;
    if api_key.is_empty() {
        bail!("PERPLEXITY_API_KEY not set in .env");
    }
    let body = serde_json::json!({
        "model": PPLX_MODEL,
        "messages": [{"role": "user", "content": query}],
    });
    let resp = reqwest::Client::new()
        .post(PPLX_URL)
        .bearer_auth(api_key)
        .json(&body)
        .timeout(std::time::Duration::from_secs(30))
        .send()
        .await?;
    let data: serde_json::Value = resp.json().await?;
    Ok(data["choices"][0]["message"]["content"]
        .as_str()
        .unwrap_or("")
        .to_string())
}
